### PR TITLE
[FIXED] Possible processing of protocols out-of-order

### LINKED ISSUE #195: Server possibly processing protocols out of order
The way the server subscribes to various protocol subjects, it is possible that protocols sent in a given order from a client are processed out of order by the server.

### PR DESCRIPTION
**Pull request by kozlovic**
All internal subscribers now use a ChanSubscription and share the
same channel. Protocols are pulled-out and form a list that is
processed in a ioLoop go routine.

Resolves #195

### CODE REVIEW

File: server/server.go [hunk 1]
```diff
@@ -1854,101 +1934,143 @@ func (s *StanServer) storeIOLoop() {
...
+			if sleepTime > 0 {
+				s.ioMutex.RLock()
+				count := s.ioList.count
+				s.ioMutex.RUnlock()
+				if count > 0 && count < batchSize {
+					time.Sleep(sleepDur)
 				}
+			}
+			// Take the list now.
+			s.ioMutex.Lock()
```
Comment (ColinSullivan1) [inline server/server.go]: For readability, this could be a function on the ioProtoList, e.g. ioProtoList.Clear()

File: server/server.go [hunk 2]
```diff
@@ -1854,101 +1934,143 @@ func (s *StanServer) storeIOLoop() {
...
+					proto.next = nil
+
+					// Process client published messages and add them to the
+					// pubMsgs list.
+					if proto.m.Sub == s.pubSub {
+						iopm, cs := s.processClientPublish(proto.m)
+						if iopm != nil {
+							proto.pubmsg = iopm
+							storesToFlush[cs] = struct{}{}
+							if pubMsgs.tail == nil {
```
Comment (ColinSullivan1) [inline server/server.go]: This could be in an ioProtoList.Append() or PushBack() function

File: server/server.go [hunk 3]
```diff
@@ -1854,101 +1934,143 @@ func (s *StanServer) storeIOLoop() {
...
-				remaining -= ioChanLen
-			}
-
-			// flush all the stores with messages written to them...
-			for cs := range storesToFlush {
-				if err := cs.Msgs.Flush(); err != nil {
-					// TODO: Attempt recovery, notify publishers of error.
-					panic(fmt.Errorf("Unable to flush msg store: %v", err))
+				for i := 0; i < limit; i++ {
+					proto := list.head
```
Comment (ColinSullivan1) [inline server/server.go]: This could be in an ioProtoList.RemoveHead() function

**Comment by wallyqs** [comment]
LGTM

**Comment by kozlovic** [comment]
@ColinSullivan1 Made the changes, not sure about the naming of some of the APIs ;-)

(note that sine `ioProtoList` is not exported, therefore the methods should not be (so start with lower case unlike in your review).

**Comment by ColinSullivan1** [comment]
LGTM.

**Pull request merged**